### DISCUSSION TITLE
Checking if we have GPU, using C library.

### DISCUSSION THREAD

**Discussion by Emreerdog** [ggml-org/llama.cpp#10625] [Q&A] [answered]
Hello, are there and procedure available for checking if the ggml will work on gpu or spotted a gpu device? 

I was using `ggml_cpu_has_cuda()` to check things until I pulled the latest version of llama.cpp. It seems that the procedure has been removed.

Thanks in advance.

**Comment by Emreerdog** [comment, marked as answer]
Thankfully I found it. It seems that when loading the model using `llama_load_model_from_file` it is iterating through devices. By this, I can see all devices:

```
// use all available devices
        for (size_t i = 0; i < ggml_backend_dev_count(); ++i) {
            ggml_backend_dev_t dev = ggml_backend_dev_get(i);
            switch (ggml_backend_dev_type(dev)) {
                case GGML_BACKEND_DEVICE_TYPE_CPU:
                case GGML_BACKEND_DEVICE_TYPE_ACCEL:
                    // skip CPU backends since they are handled separately
                    break;

                case GGML_BACKEND_DEVICE_TYPE_GPU:
                    model->devices.push_back(dev);
                    break;
            }
        }
```

**Reply by slaren** [reply]
This is indeed the recommended way to deal with this. Do not rely on compile time settings, and use the backend registry instead. Note that backends can be loaded dynamically now, so there is no guarantee that the backends included in the build are the ones available. llama.cpp has already been updated to work in this way. Alternatively, to ease the transition with old code, you can also check if the macro `GGML_USE_CUDA` is defined.
